src/peft/tuners/xlora/model.py

```diff
@@ -0,0 +1,383 @@
+                target=module,
+                target_forward=module.forward,
+                layer_number=total_swapped,
+                config=config,
+            )
+            module.forward = new_layer.forward  # type: ignore[method-assign]
+            total_swapped += 1
+        elif isinstance(module, lora.Conv2d):
+            device = module.lora_A[next(iter(module.lora_A))].weight.device
+            new_layer = XLoRAConv2dLayer(
+                model=xloramodel,
+                target=module,
+                target_forward=module.forward,
+                layer_number=total_swapped,
+                config=config,
+            )
+            module.forward = new_layer.forward  # type: ignore[method-assign]
+            total_swapped += 1
+
+    return (total_swapped, device)
+
+
+class XLoraModel(BaseTuner):
+    """
+    Creates an X-LoRA (Mixture of LoRA experts), model from a pretrained transformers model. Currently,
+    this X-LoRA implementation only works with models with a transformer architecture.
+
+    The method is described in detail in https://arxiv.org/abs/2402.07148.
+
+    Args:
+        model ([`torch.nn.Module`]): The model to be adapted.
+        config ([`XLoraConfig`]): The configuration of the Lora model.
+        adapter_name (`str`): The name of the adapter, does not affect the LoRA adapter names.
+
+    Returns:
+        `torch.nn.Module`: The X-LoRA model.
+
+    Example:
+        ```py
+        >>> from transformers import AutoModelForCausalLM, AutoConfig
+        >>> from peft import LoraConfig, PeftModel, get_peft_model, prepare_model_for_int8_training
+
+        >>> model_config = AutoConfig.from_pretrained("mistralai/Mistral-7B-Instruct-v0.1")
+        >>> config = XLoraConfig(
+        ...     task_type="CAUSAL_LM",
+        ...     hidden_size=model_config.hidden_size,
+        ...     xlora_depth=4,
+        ...     adapters={
+        ...         "adapter_1": "./path/to/the/checkpoint/",
+        ...         "adapter_2": "./path/to/the/checkpoint/",
+        ...         "adapter_n": "./path/to/the/checkpoint/",
+        ...     },
+        ... )
+
+        >>> model = AutoModelForCausalLM.from_pretrained(
+        ...     "mistralai/Mistral-7B-Instruct-v0.1",
+        ...     trust_remote_code=True,
+        ...     use_flash_attention_2=False,
+        ...     device_map="cuda:0",
+        ...     torch_dtype=torch.bfloat16,
+        ... )
+        >>> model = prepare_model_for_int8_training(model)
+        >>> xlora_model = get_peft_model(model, config)
+        ```
+    """
+
+    def __init__(
+        self,
+        model: nn.Module,
+        config: Union[dict[str, XLoraConfig], XLoraConfig],
+        adapter_name: str,
+    ) -> None:
+        if isinstance(config, dict):
+            conf = config[adapter_name]
+        else:
+            conf = config
+        lora_model = LoraModel(model, config.copy(), adapter_name)
+        self.__dict__["xlora_config"] = conf
```

Comment:
My understanding is that you go through `self.__dict__` to avoid triggering the registration process that `nn.Module` performs when adding certain attributes, maybe because of circular references or something like that. But I would like to understand what the issue is specifically. My feeling is that we should solve that issue differently, maybe by restructuring the code, since this type of "hack" can lead to bugs that are hard to track down.

Reply:
Yes, it is certainly a hack. I'm not sure of a restructure that we can do to fix this, as we need the `self.lora_model` and `self.xlora_config` attributes to be accessible before `super().__init__(model, config, adapter_name)` because it calls into the XLoraModel. I'll take a look and see if I can restructure it.